network_simulator/flow.py

```diff
@@ -125,38 +173,428 @@ def handle_packet_loss(self, packet_id, loss_type, main_event_loop):
+        packets_in_transit_copy.add(packet_id)
+
+        if len(packets_in_transit_copy) > int(self.window_size_packets):
+            logger.debug("Flow %s could not retransmit packet %d since its "
+                         "window size was too small. Will add packet to "
+                         "buffer.")
+            self.packet_id_buffer.put_nowait(packet_id)
+            return False
+        else:
+            this_packet = DataPacket(packet_id=packet_id,
+                                     flow_id=self.flow_id,
+                                     source_id=self.source_addr,
+                                     dest_id=self.dest_addr,
+                                     start_time_sec=
+                                     main_event_loop.global_clock_sec)
+            flow_send_event = FlowSendPacketsEvent(self, [this_packet])
+            main_event_loop.schedule_event_with_delay(flow_send_event, 0.0)
+            logger.debug("Flow %s successfully scheduled a retransmit of "
+                         "packet %d.")
+            return True
+
+    def send_packets_to_fill_window(self, main_event_loop):
+        """
+        A helper function that sends enough DataPackets to fill the current
+        window of the TCP algorithm.
+        :param MainEventLoop main_event_loop: event loop for scheduling Events.
+        """
+        curr_window_size = len(self.packets_in_transit)
+        packets_to_send = []
+        while curr_window_size < int(self.window_size_packets):
+            new_packet_id = self.get_next_data_packet_id()
+            if self.packet_id_exceeds_data(new_packet_id):
+                # We are out of Packets to send. Send what's already there.
+                break
+
+            new_packet = DataPacket(packet_id=new_packet_id,
+                                    flow_id=self.flow_id,
+                                    source_id=self.source_addr,
+                                    dest_id=self.dest_addr,
+                                    start_time_sec=
+                                    main_event_loop.global_clock_sec)
+            packets_to_send.append(new_packet)
+            curr_window_size += 1
+
+        if len(packets_to_send) > 0:
+            flow_send_event = FlowSendPacketsEvent(self, packets_to_send)
+            main_event_loop.schedule_event_with_delay(flow_send_event, 0.0)
 
     def __repr__(self):
         return str(self.__dict__)
 
+    def __eq__(self, other):
+        return isinstance(other, self.__class__) and \
+               self.flow_id == other.flow_id
+
+
+class FlowDummy(Flow):
+    """
+    A dummy Flow implementation, for testing only. This just maintains a
+    window size of 1 packet. Successes and losses do not change the window
+    size. Losses result in retransmits.
+    """
+
+    def __init__(self, flow_id=None, source_addr=None, dest_addr=None,
+                 source=None, dest=None,
+                 data_size_bits=None, start_time_sec=None):
```

Comment:
In subclasses, I only add :ivars for instance variables that were not in Flow. There aren't any in this case.